api.go

```diff
@@ -177,6 +177,66 @@ func (r *lockedRandSource) Seed(seed int64) {
 	r.lk.Unlock()
 }
 
+// Redirect requests by re signing the request.
+func (c *Client) redirectHeaders(req *http.Request, via []*http.Request) error {
+	if len(via) >= 5 {
+		return errors.New("stopped after 5 redirects")
+	}
+	if len(via) == 0 {
+		return nil
+	}
+	lastRequest := via[len(via)-1]
+	var reAuth bool
+	for attr, val := range lastRequest.Header {
+		// if hosts do not match do not copy Authorization header
+		if attr == "Authorization" && req.Host != lastRequest.Host {
+			reAuth = true
+			continue
+		}
+		if _, ok := req.Header[attr]; !ok {
+			req.Header[attr] = val
+		}
+	}
+	c.endpointURL = *req.URL
+	value, err := c.credsProvider.Get()
+	if err != nil {
+		return err
+	}
+	var (
+		signerType      = value.SignerType
+		accessKeyID     = value.AccessKeyID
+		secretAccessKey = value.SecretAccessKey
+		sessionToken    = value.SessionToken
+		region          = c.region
+	)
+
+	// Custom signer set then override the behavior.
+	if c.overrideSignerType != credentials.SignatureDefault {
+		signerType = c.overrideSignerType
+	}
+
+	// If signerType returned by credentials helper is anonymous,
+	// then do not sign regardless of signerType override.
+	if value.SignerType == credentials.SignatureAnonymous {
+		signerType = credentials.SignatureAnonymous
+	}
+
+	if reAuth {
+		// Check if there is no region override, if not get it from the URL if possible.
+		if region == "" {
+			region = getRegionFromURL(c.endpointURL)
+		}
+		switch {
+		case signerType.IsV2():
+			// Add signature version '2' authorization header.
+			req = s3signer.SignV2(*req, accessKeyID, secretAccessKey)
+		case signerType.IsV4():
+			req = s3signer.SignV4(*req, accessKeyID, secretAccessKey, sessionToken, getDefaultLocation(c.endpointURL, region))
```

Comment:
This is using `c.endpointURL`.  Curious as to why not use the redirected host?

Reply:
c.endpointURL is replaced with the new host.

Reply:
So all future operations happen directly on new host rather than doing a continuous redirection for all requests.

Reply:
Ah, this make sense.  Thanks!  But this leads me to another question-

Line 51, `endpointURL` isn't a pointer.  And `executeMethod` on 525 doesn't have a pointer receiver.  So, for every request, `executeMethod` will get the redirect, the go client will retry and follow the `Location`, we end up here and resign the request after setting the new `endpoint`.  

Basically changes to `c.endpointURL` won't persist between requests, right?  The redirect will happen at least once per request OR the client will need to be updated.

Reply:
Not really. *Client is a pointer so endpointURL gets updated anyways, you can test this out yourself.

Reply:
That is one of the reasons why redirectHeaders() function has a pointer receiver.

Reply:
:+1:   yup, you're right.  Thanks for indulging me.  This patch works for my use-case.  Tested it with a redirector without an issue.

Reply:
For my own edification, I did some probing.

For every request, there is 1 redirection.  But subsequent calls also recieve a redirection because `endpoint` is not updated.  And this is because because `do` does not have a pointer receiver.  Infact, every call in the stack before `do` doesn't either.  Notice the lack of `(*Client)` in calls before the frame with `redirectHeaders` in the stack below.

```
goroutine 1 [running]:
github.com/minio/mc/vendor/github.com/minio/minio-go.(*Client).redirectHeaders(0xc4201f8a20, 0xc420202300, 0xc42000e228, 0x1, 0x1, 0xc42008ff10, 0xc4201e5590)
	/food/s4/src/github.com/minio/mc/vendor/github.com/minio/minio-go/api.go:182 +0x39
github.com/minio/mc/vendor/github.com/minio/minio-go.(*Client).(github.com/minio/mc/vendor/github.com/minio/minio-go.redirectHeaders)-fm(0xc420202300, 0xc42000e228, 0x1, 0x1, 0xc4201e6a50, 0x25)
	/food/s4/src/github.com/minio/mc/vendor/github.com/minio/minio-go/api.go:269 +0x52
net/http.(*Client).checkRedirect(0xc4201e5170, 0xc420202300, 0xc42000e228, 0x1, 0x1, 0x0, 0x101)
	/usr/local/go/src/net/http/client.go:411 +0x5a
net/http.(*Client).Do(0xc4201e5170, 0xc420202200, 0xc, 0x0, 0x0)
	/usr/local/go/src/net/http/client.go:568 +0x918
github.com/minio/mc/vendor/github.com/minio/minio-go.Client.do(0xc4201fa6c0, 0x4, 0x0, 0x0, 0x0, 0xc4201fa6c7, 0xf, 0x0, 0x0, 0x0, ...)
	/food/s4/src/github.com/minio/mc/vendor/github.com/minio/minio-go/api.go:474 +0x50
github.com/minio/mc/vendor/github.com/minio/minio-go.Client.getBucketLocation(0xc4201fa6c0, 0x4, 0x0, 0x0, 0x0, 0xc4201fa6c7, 0xf, 0x0, 0x0, 0x0, ...)
	/food/s4/src/github.com/minio/mc/vendor/github.com/minio/minio-go/bucket-cache.go:104 +0x1a4
<snip>
```

And my trivial redirector confirms this.  I issued an `mc ls redirectedHost/foo`
```
redirecting from /foo/?location=
redirecting from /foo/
```
If the endpoint in the client was updated correctly, we should have seen only 1 redirect.

Simplified example
https://play.golang.org/p/TTKjoFO_Wz

Anyway, again this patch does work so the above is a minor gripe.  But if you reuse the client, you'll see 1 redirect for each request.  Maybe that's desireable.

Reply:
Yes let me verify it again. And fix it properly making pointer receiver is not a big deal